src/app.rs

```diff
@@ -28,59 +23,81 @@ pub struct App<Model, Message, Init, Update, View> {
     update: Rc<Update>,
     view: Rc<View>,
     root_id: String,
-    state: Rc<RefCell<State<Model, Message>>>,
+    state: Rc<RefCell<Option<State<Model, Message>>>>,
 }
 
 impl<Model, Message, Init, Update, View> App<Model, Message, Init, Update, View>
 where
     Model: 'static + Clone + fmt::Debug + Eq,
     Message: 'static + Clone + fmt::Debug,
-    Init: 'static + Fn() -> Model,
-    Update: 'static + Fn(&Message, &Model) -> Model,
+    Init: 'static + Fn() -> (Model, Commands<Message>),
+    Update: 'static + Fn(&Message, &Model) -> (Model, Commands<Message>),
     View: 'static + Fn(&Model) -> Html<Message>,
 {
     pub fn new(init: Init, update: Update, view: View, root_id: &str) -> Self {
-        let state = State::initial(init());
-
         Self {
             init: Rc::new(init),
             update: Rc::new(update),
             view: Rc::new(view),
             root_id: root_id.into(),
-            state: Rc::new(RefCell::new(state)),
+            state: Rc::new(RefCell::new(None)),
         }
     }
 
     pub fn start(&self) {
-        let model = self.state().model.clone();
-        let html = (self.view)(&model);
-        self.render_app(&html).unwrap();
-        self.set_state(State::new(model, html));
+        let (model, commands) = (self.init)();
+
+        self.update(model, commands);
     }
 
     pub fn handle_message(&self, message: &Message) {
-        let new_model = (self.update)(message, &self.state().model);
-        if new_model == self.state().model {
+        info!("message: {:#?}", message);
+
+        let (new_model, commands) = (self.update)(message, &self.state().as_ref().unwrap().model);
+
+        if new_model == self.state().as_ref().unwrap().model {
             return;
         }
 
-        let new_html = (self.view)(&new_model);
+        self.update(new_model, commands);
+    }
+
+    fn update(&self, model: Model, commands: Commands<Message>) {
+        info!("model: {:#?}", model);
+
+        let new_html = (self.view)(&model);
         self.render_app(&new_html).unwrap();
 
-        self.set_state(State::new(new_model, new_html));
+        self.handle_commands(commands);
+        self.set_state(State::new(model, new_html));
+    }
+
+    fn handle_commands(&self, commands: Commands<Message>) {
+        for command in commands {
+            let app = self.clone();
+
+            spawn_local(async move {
+                let message = command.run().await;
+                app.handle_message(&message);
+            });
```

Comment:
wasm_bindgen_futures provides an async runtime for WASM, with interop between Rust Futures and JS Promises. `spawn_local` is roughly equivalent to spawning off a thread.